Aesara fails to set the correct default blas ldflags to link with MKL on windows

## Description of your problem or feature request

MKL libraries on windows are supplied as dynamic link libraries (e.g. `mkl_core.1.dll`) and don't have the lib prefix. When aesara tries to set the default blas ldflags [here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L2635) it does a few things:

1. It tries to import `mkl`, which should be available if `mkl-services` was installed via conda or mamba.
2. Sets up a gcc styled library directory flag `-L` [see here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L2763)
3. It sets the default libraries to try to link: `-lmkl_core -lmkl_rt -lmkl_intel_thread`
4. It then tries to compile and run a small [test script that links to mkl](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L1961-L1993)


**Please provide the full traceback of any errors.**
The test script fails to compile and run [here](https://github.com/aesara-devs/aesara/blob/f7a506ff3a599a8101502b7ff4d7b0c642366a22/aesara/link/c/cmodule.py#L1985-L1987). You can run it manually using this snippet:

```python
import textwrap
from aesara.link.c.cmodule import std_lib_dirs, GCC_compiler
import os
import sys


flags = [
    '-L"C:\\Users\\chulo\\Anaconda3\\Library\\bin"',
    '-lmkl_core',
    '-lmkl_intel_thread',
    '-lmkl_rt',
]


test_code = textwrap.dedent(
    """\
    extern "C" double ddot_(int*, double*, int*, double*, int*);
    int main(int argc, char** argv)
    {
        int Nx = 5;
        int Sx = 1;
        double x[5] = {0, 1, 2, 3, 4};
        double r = ddot_(&Nx, x, &Sx, x, &Sx);
        if ((r - 30.) > 1e-6 || (r - 30.) < -1e-6)
        {
            return -1;
        }
        return 0;
    }
    """
)
cflags = list(flags)
# to support path that includes spaces, we need to wrap it with double quotes on Windows
path_wrapper = '"' if os.name == "nt" else ""
cflags.extend([f"-L{path_wrapper}{d}{path_wrapper}" for d in std_lib_dirs()])

res = GCC_compiler.try_compile_tmp(
    test_code, tmp_prefix="try_blas_", flags=cflags, try_run=True, output=True
)
```

This results in
```python
>>> print(res[-1].decode("utf-8"))

C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_core: No such file or directory
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_intel_thread: No such file or directory
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/11.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -lmkl_rt: No such file or directory
collect2.exe: error: ld returned 1 exit status
```


**Please provide any additional information below.**

Since the mkl .dll libraries don't have the lib prefix, we should provide their absolute paths to be able to link them when compiling. This means that setting the flags to:

```python
import sys

flags = [
    os.path.join(sys.prefix, "Library", "bin", "mkl_core.1.dll"),
    os.path.join(sys.prefix, "Library", "bin", "mkl_intel_thread.1.dll"),
    os.path.join(sys.prefix, "Library", "bin", "mkl_rt.1.dll"),
]
```

makes the compilation succeed.

## Versions and main components

* Aesara version: 2.6.2
* Aesara config (`python -c "import aesara; print(aesara.config)"`)
* Python version: 3.9.7
* Operating system: Windows 10
* How did you install Aesara: (conda/pip) conda

--- a/aesara/link/c/cmodule.py
+++ b/aesara/link/c/cmodule.py
@@ -14,6 +14,7 @@ import shutil
 import stat
 import subprocess
 import sys
+import sysconfig
 import tempfile
 import textwrap
 import time
@@ -1678,7 +1679,7 @@ def std_lib_dirs_and_libs() -> Optional[Tuple[List[str], ...]]:
         # Obtain the library name from the Python version instead of the
         # installation directory, in case the user defined a custom
         # installation directory.
-        python_version = distutils.sysconfig.get_python_version()
+        python_version = sysconfig.get_python_version()
         libname = "python" + python_version.replace(".", "")
         # Also add directory containing the Python library to the library
         # directories.
@@ -2381,7 +2382,13 @@ class GCC_compiler(Compiler):
         comp_args=True,
     ):
         return cls._try_compile_tmp(
-            src_code, tmp_prefix, flags, try_run, output, config.cxx, comp_args
+            src_code,
+            tmp_prefix,
+            cls.patch_ldflags(flags),
+            try_run,
+            output,
+            config.cxx,
+            comp_args,
         )
 
     @classmethod
@@ -2395,8 +2402,57 @@ class GCC_compiler(Compiler):
         comp_args=True,
     ):
         return cls._try_flags(
-            flag_list, preamble, body, try_run, output, config.cxx, comp_args
+            cls.patch_ldflags(flag_list),
+            preamble,
+            body,
+            try_run,
+            output,
+            config.cxx,
+            comp_args,
         )
+
+    @staticmethod
+    def patch_ldflags(flag_list: List[str]) -> List[str]:
+        lib_dirs = [flag[2:].lstrip() for flag in flag_list if flag.startswith("-L")]
+        flag_idxs: List[int] = []
+        libs: List[str] = []
+        for i, flag in enumerate(flag_list):
+            if flag.startswith("-l"):
+                flag_idxs.append(i)
+                libs.append(flag[2:].lstrip())
+        if not libs:
+            return flag_list
+        libs = GCC_compiler.linking_patch(lib_dirs, libs)
+        for flag_idx, lib in zip(flag_idxs, libs):
+            flag_list[flag_idx] = lib
+        return flag_list
+
+    @staticmethod
+    def linking_patch(lib_dirs: List[str], libs: List[str]) -> List[str]:
+        if sys.platform != "win32":
+            return [f"-l{l}" for l in libs]
+
+        def sort_key(lib):  # type: ignore
+            name, *numbers, extension = lib.split(".")
+            return (extension == "dll", tuple(map(int, numbers)))
+
+        patched_lib_ldflags = []
+        for lib in libs:
+            ldflag = f"-l{lib}"
+            for lib_dir in lib_dirs:
+                lib_dir = lib_dir.strip('"')
+                windows_styled_libs = [
+                    fname
+                    for fname in os.listdir(lib_dir)
+                    if not (os.path.isdir(os.path.join(lib_dir, fname)))
+                    and fname.split(".")[0] == lib
+                    and fname.split(".")[-1] in ["dll", "lib"]
+                ]
+                if windows_styled_libs:
+                    selected_lib = sorted(windows_styled_libs, key=sort_key)[-1]
+                    ldflag = f'"{os.path.join(lib_dir, selected_lib)}"'
+            patched_lib_ldflags.append(ldflag)
+        return patched_lib_ldflags
 
     @staticmethod
     def compile_str(
@@ -2509,7 +2565,7 @@ class GCC_compiler(Compiler):
             cmd.append("-fvisibility=hidden")
         cmd.extend(["-o", f"{path_wrapper}{lib_filename}{path_wrapper}"])
         cmd.append(f"{path_wrapper}{cppfilename}{path_wrapper}")
-        cmd.extend([f"-l{l}" for l in libs])
+        cmd.extend(GCC_compiler.linking_patch(lib_dirs, libs))
         # print >> sys.stderr, 'COMPILING W CMD', cmd
         _logger.debug(f"Running cmd: {' '.join(cmd)}")
 

--- a/tests/link/c/test_cmodule.py
+++ b/tests/link/c/test_cmodule.py
@@ -5,6 +5,7 @@ But this one tests a current behavior that isn't good: the c_code isn't
 deterministic based on the input type and the op.
 """
 import logging
+import os
 import tempfile
 from unittest.mock import patch
 
@@ -91,3 +92,50 @@ def test_default_blas_ldflags(sys_mock, try_blas_flag_mock, caplog):
             default_blas_ldflags()
 
     assert caplog.text == ""
+
+
+@patch(
+    "os.listdir", return_value=["mkl_core.1.dll", "mkl_rt.1.0.dll", "mkl_rt.1.1.lib"]
+)
+@patch("sys.platform", "win32")
+def test_patch_ldflags(listdir_mock):
+    mkl_path = "some_path"
+    flag_list = ["-lm", "-lopenblas", f"-L {mkl_path}", "-l mkl_core", "-lmkl_rt"]
+    assert GCC_compiler.patch_ldflags(flag_list) == [
+        "-lm",
+        "-lopenblas",
+        f"-L {mkl_path}",
+        '"' + os.path.join(mkl_path, "mkl_core.1.dll") + '"',
+        '"' + os.path.join(mkl_path, "mkl_rt.1.0.dll") + '"',
+    ]
+
+
+@patch(
+    "os.listdir",
+    return_value=[
+        "libopenblas.so",
+        "libm.a",
+        "mkl_core.1.dll",
+        "mkl_rt.1.0.dll",
+        "mkl_rt.1.1.dll",
+    ],
+)
+@pytest.mark.parametrize("platform", ["win32", "linux", "darwin"])
+def test_linking_patch(listdir_mock, platform):
+    libs = ["openblas", "m", "mkl_core", "mkl_rt"]
+    lib_dirs = ['"mock_dir"']
+    with patch("sys.platform", platform):
+        if platform == "win32":
+            assert GCC_compiler.linking_patch(lib_dirs, libs) == [
+                "-lopenblas",
+                "-lm",
+                '"' + os.path.join(lib_dirs[0].strip('"'), "mkl_core.1.dll") + '"',
+                '"' + os.path.join(lib_dirs[0].strip('"'), "mkl_rt.1.1.dll") + '"',
+            ]
+        else:
+            GCC_compiler.linking_patch(lib_dirs, libs) == [
+                "-lopenblas",
+                "-lm",
+                "-lmkl_core",
+                "-lmkl_rt",
+            ]

